### PR TITLE
Plugins folder with example, build config, and template

### PR DESCRIPTION
**Pull request by illume**
- [x] make a plugins folder
- [x] headlamp-plugin: Move plugins-pkg plugins/headlamp-plugin
- [x] headlamp-plugin: Copy compiled files to config dir
- [x] headlamp-plugin: Add a --watch for development
- [x] plugins: Add headlamp-pod-counter example
- [x] plugins: Add README.md files
- [x] plugins: Add plugins template
- [x] headlamp-plugin: Add init command and new plugin template
- [x] docs: Update plugins docs with how to use template
- [x] headlamp-plugin: Add better cmdline including help
- [x] backend: Move loading the plugins into the plugins module
- [x] backend: Reload list of plugins to serve when they're changed
- [x] headlamp-plugin: Add extract command to copy plugins
- [x] backend: Only consider existing plugin paths when listing plugins 

# How to use

Link the headlamp-plugin locally to test it before new @kinvolk/headlamp-plugin version.
```bash
cd plugins/headlamp-plugin/ && npm link
```

See the help:
```bash
npx @kinvolk/headlamp-plugin --help
npx @kinvolk/headlamp-plugin # --help is also the default when no arguments are supplied
```

This is how to start a new plugin:
```bash
npx @kinvolk/headlamp-plugin create headlamp-myfancy
cd headlamp-myfancy
npm run start
```

This is how to extract dist/main.js plugins into the pluginName/main.js form:
```bash
npx @kinvolk/headlamp-plugin extract /tmp/plugin-packages/ .plugins
```

This starts watching for plugin changes (used by the plugin start command):
```bash
npx @kinvolk/headlamp-plugin start
```

# Testing done

- [x] extract command works on folder of packages and single packages
- [x] headlamp-plugin create makes a new folder with the package name set
- [x] the created package start and build commands work
- [x] Headlamp finds new plugins put into the folder after first load.
- [x] Headlamp reloads when plugins change (when run in dev mode via make frontend Ubuntu).
- [x] Headlamp reloads when plugins change (when run as an app Ubuntu).
- [x] Headlamp reloads when plugins change (when run in dev mode via make frontend Mac).
- [x] Headlamp reloads when plugins change (when run as an app Mac).
- [x] examples/plugins/pod-counter build and start command work, and plugin shows up in Headlamp app bar.

### CODE REVIEW

File: plugins/headlamp-plugin/bin/headlamp-plugin.js [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+      path,
+      fs.readFileSync(path, 'utf8').split('$${name}').join(name)
+    )
+  }
+
+  replaceFileVariables(packagePath);
+  replaceFileVariables(indexPath);
+
+  // @todo: this is to make testing locally easier. Remove before merge.
+  const proc1 = child_process.spawnSync('npm', ['link', '@kinvolk/headlamp-plugin'], {cwd: dstFolder});
```
Comment (illume) [inline plugins/headlamp-plugin/bin/headlamp-plugin.js:55]: This needs to be removed before merging.

File: plugins/headlamp-pod-counter/package.json [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    "kubernetes",
+    "plugins"
+  ],
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/kinvolk/headlamp.git",
+    "directory": "plugins/headlamp-plugin-counter"
+  },
+  "devDependencies": {
+    "@kinvolk/headlamp-plugin": "file:../headlamp-plugin"
```
Comment (illume) [inline plugins/headlamp-pod-counter/package.json:23]: This needs updating before merging to @kinvolk/headlamp-plugin 0.2.0 once the package has been published.

File: plugins/headlamp-pod-counter/.gitignore [hunk 3]
```diff
@@ -0,0 +1,23 @@
+# See https://help.github.com/articles/ignoring-files/ for more about ignoring files.
```
Comment (joaquimrocha) [inline plugins/headlamp-pod-counter/.gitignore:1]: I think we shouldn't mix plugins with the `headlamp-plugin` pkg at the same level, and we're also not intending yet to ship any official plugins in the repo. Therefore, I'd rather have the structure as:
```
plugins/
  README.md
  headlamp-plugin/
  examples/
    pod-counter/
    ...
```

File: plugins/headlamp-plugin/bin/headlamp-plugin.js [hunk 4]
```diff
@@ -6,6 +6,39 @@ const webpack = require('webpack');
...
+        onEnd: {
+          copy: [{
+            source: './dist/*',
+            destination: `${paths.config}/plugins/${packageName}/`,
+          }],
+        },
+      }
+    }),
+  ];
+}
```
Comment (joaquimrocha) [inline plugins/headlamp-plugin/bin/headlamp-plugin.js:95]: Would it work if we just used a symlink? (not sure if there are implications on Windows)
The advantage would be that it'd be faster, and if we also name it with a known prefix (e.g. `__dev_`+plugin_name) in order to clean up broken links (once development is finished).

File: plugins/headlamp-plugin/template/package.json [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+  "name": "$${name}",
+  "version": "0.0.1",
+  "description": "Your Headlamp plugin",
+  "scripts": {
+    "start": "headlamp-plugin --watch",
+    "build": "headlamp-plugin"
+  },
+  "keywords": [
+    "headlamp",
+    "kinvolk",
```
Comment (joaquimrocha) [inline plugins/headlamp-plugin/template/package.json:11]: Please remove `kinvolk` as I am not sure it should be associated with all plugins people develop.

File: docs/development/plugins/building.md [hunk 6]
```diff
@@ -3,57 +3,16 @@ title: Building and Shipping Plugins
...
-      index.tsx
-    package.json
-  ```
-
-Feel free to use `npm create` for getting the base structure/files right.
-
-For helping with compiling the plugin code we have the
-[@kinvolk/headlamp-plugin](https://www.npmjs.com/package/@kinvolk/headlamp-plugin),
-which should be installed as a dev dependency:
+This is how to start a new plugin:
```
Comment (joaquimrocha) [inline docs/development/plugins/building.md:6]: You have WIP in the title, not sure if it was forgotten.

File: plugins/headlamp-plugin/README.md [hunk 7]
```diff
@@ -3,3 +3,12 @@
...
 
 Headlamp plugins depend on this package.
+
+
+## Commands
+
+```
+headlamp-plugin build        Build the plugin                     [default]
+headlamp-plugin watch        Watch for changes and build the plugin
+headlamp-plugin init <name>  Initialize a new plugin folder with base code
```
Comment (joaquimrocha) [inline plugins/headlamp-plugin/README.md:13]: Please note the comment about this name.

File: plugins/headlamp-plugin/bin/headlamp-plugin.js [hunk 8]
```diff
@@ -125,16 +122,27 @@ function compile(err, stats) {
...
+    webpack(config, compile);
+  })
+  .command('watch', 'Watch for changes and build the plugin', {}, (argv) => {
+    config.watch = true;
+    config.mode = 'development';
+    process.env['BABEL_ENV'] = 'development';
+    copyToPluginsFolder(config);
+    webpack(config, compile);
+  })
+  .command('init <name>', 'Initialize a new plugin folder with base code', (yargs) => {
```
Comment (joaquimrocha) [inline plugins/headlamp-plugin/bin/headlamp-plugin.js:139]: I think from a product point of view, doing `headlamp-plugin create NAME` sounds better. Init is more ambiguous.

**Review by joaquimrocha — COMMENTED** [review]
Left a couple of comments, but overall it's looking good.

File: plugins/headlamp-plugin/bin/headlamp-plugin.js (hunk 4 above)
Comment (illume) [reply plugins/headlamp-plugin/bin/headlamp-plugin.js:95]: I'm pretty sure you need some special permissions on windows for symlinks. I had to do this recently, and required to run as admin.

**Comment by joaquimrocha** [comment]
@illume I have added 2 commits with a rough version of hot reloading the list of plugins.
I am hitting CORS issues with it (even if using `-dev`), so this is WIP. I think it's important for the server to update its list of plugins if a plugin is deleted or added, as part of the overall plugin dev experience.

**Comment by joaquimrocha** [comment]
Forgot to add that maybe we should change the logic a bit into an event driven one, connecting functions to the plugins' events (so we can reload the plugin list and send the reload signal to the UI independently from each other). But this can be done at a later point.

File: plugins/headlamp-plugin/template/package.json (hunk 5 above)
Comment (illume) [reply plugins/headlamp-plugin/template/package.json]: This is done.

File: plugins/headlamp-plugin/bin/headlamp-plugin.js (hunk 8 above)
Comment (illume) [reply plugins/headlamp-plugin/bin/headlamp-plugin.js:139]: This is changed to create.

File: docs/development/plugins/building.md (hunk 6 above)
Comment (illume) [reply docs/development/plugins/building.md:6]: The title was fixed.

File: plugins/headlamp-pod-counter/.gitignore (hunk 3 above)
Comment (illume) [reply plugins/headlamp-pod-counter/.gitignore]: Yeah, this is better. It's this way now.

**Comment by joaquimrocha** [comment]
I have made the plugin list function a bit more robust. The plugins seem to be updated correctly now.
I still have trouble with the `watch` command (now it's not working for me). Maybe it's got to do with the limit on the number of watched files.

**Comment by illume** [comment]
I added the `headlamp-plugin extract` command for extracting plugins into our pluginName/main.js form. I have a change to it so that it automatically creates the output path if it doesn't exist.

Also made --help the default rather than `build`, breaking backwards compatibility (so lokomotive web ui will need an update).

File: backend/cmd/plugins.go [hunk 9]
```diff
@@ -149,7 +149,15 @@ func (c *HeadlampConfig) getPluginListURLs() ([]string, error) {
 	pluginListURLs := make([]string, 0, len(files))
 
 	for _, f := range files {
-		if f.Name() == ".gitignore" {
+		if !f.IsDir() {
+			continue
+		}
+
+		pluginPath := filepath.Join(c.pluginDir, f.Name(), "main.js")
```
Comment (illume) [inline backend/cmd/plugins.go:156]: :+1:

**Comment by illume** [comment]
The docs commit #42af94d looks good to me, and I removed the debugging code. 

ps. The plugins/headlamp-plugins package needs to be published at some point soon too.

File: plugins/headlamp-plugin/bin/headlamp-plugin.js [hunk 10]
```diff
@@ -51,8 +51,8 @@ function init(name) {
   replaceFileVariables(packagePath);
   replaceFileVariables(indexPath);
 
-  // @todo: this is to make testing locally easier. Remove before merge.
-  const proc1 = child_process.spawnSync('npm', ['link', '@kinvolk/headlamp-plugin'], {cwd: dstFolder});
+  // This can be used to make testing locally easier.
+  // const proc1 = child_process.spawnSync('npm', ['link', '@kinvolk/headlamp-plugin'], {cwd: dstFolder});
```
Comment (joaquimrocha) [inline plugins/headlamp-plugin/bin/headlamp-plugin.js:55]: Don't we just have to run `npm link` in the `headlamp-plugin` folder? If that's all, I think we should remove this commented out code.

File: plugins/headlamp-plugin/bin/headlamp-plugin.js [hunk 11]
```diff
@@ -51,6 +52,8 @@ function create(name) {
   replaceFileVariables(packagePath);
   replaceFileVariables(indexPath);
 
+  console.log(`Installing dependencies...`);
```
Comment (joaquimrocha) [inline plugins/headlamp-plugin/bin/headlamp-plugin.js:55]: Nitpick: It's using backticks but uses no code within.

**Comment by illume** [comment]
todo: On mac, `headlamp-plugin start` is using "~/Library/Preferences/Headlamp/plugins/" and should be using "~/Library/Application Support/Headlamp/plugins"

**Pull request merged**